R/growth_rate.R

```diff
@@ -195,71 +214,179 @@ growth_rate <- function(x = seq_along(y), y, x0 = x,
         }
       }
     })
-
-    return(g[i0])
+    return(stats::approx(x, g, x0)$y)
   }
 
   # Global methods
   if (method == "smooth_spline" || method == "trend_filter") {
-    # Convert to numerics
-    x <- as.numeric(x)
-    x0 <- as.numeric(x0)
-
-    # Collect parameters
-    params <- list(...)
+    if (any(is.na(x) | is.na(y) | !is.finite(x) | !is.finite(y))) {
+      cli_abort(c(
+        "{.val {method}} requires all real values without missingness.",
+        i = "Set `na_rm = TRUE` and / or check for infinite values.",
+        i = "Using `log_scale = TRUE` may induce either case."
+      ))
+    }
 
-    # Smoothing spline
     if (method == "smooth_spline") {
-      params$x <- x
-      params$y <- y
-      obj <- do.call(stats::smooth.spline, params)
+      if (is.character(params$df)) params$df <- NULL
+      if (length(params$lambda) > 1L) {
+        cli_abort("{.val smooth_spline} requires 1 `lambda` but more were used.")
+      }
+      params <- params[c("df", "spar", "lambda", "cv", "all.knots", "df.offset", "penalty")]
+      params <- params[!sapply(params, is.null)]
+      obj <- rlang::inject(stats::smooth.spline(x = x, y = y, !!!params))
       f0 <- stats::predict(obj, x = x0)$y
       d0 <- stats::predict(obj, x = x0, deriv = 1)$y
       if (log_scale) {
         return(d0)
       } else {
         return(d0 / f0)
       }
-    } else {
-      # Trend filtering
-      ord <- params$ord
-      maxsteps <- params$maxsteps
-      cv <- params$cv
-      df <- params$df
-      k <- params$k
-
-      # Default parameters
-      ord <- ord %||% 3
-      maxsteps <- maxsteps %||% 1000
-      cv <- cv %||% TRUE
-      df <- df %||% "min"
-      k <- k %||% 3
-
-      # Check cv and df combo
-      if (is.numeric(df)) cv <- FALSE
-      if (!cv && !(is.numeric(df) && df == round(df))) {
-        cli_abort("If `cv = FALSE`, then `df` must be an integer.")
-      }
-
-      # Compute trend filtering path
-      obj <- genlasso::trendfilter(y = y, pos = x, ord = ord, max = maxsteps)
-
-      # Use CV to find df, if we need to
-      if (cv) {
-        cv_obj <- quiet(genlasso::cv.trendfilter(obj, k = k, mode = "df"))
-        df <- ifelse(df == "min", cv_obj$df.min, cv_obj$df.1se)
+    } else { # Trend filtering
+      params <- parse_trendfilter_params(params)
+      sdy <- stats::sd(y) # speeds computation significantly
+      y <- y / sdy
```

Comment:
question: why only scale not shift+scale?  If scale only, should it be ~ sqrt(mean(y^2))?